Fix_kql_issue_found_by_wingfuzz

kashwy opened the pull request:
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix segmentation fault in KQL parser when the input query exceeds the `max_query_size`. Also re-enable the KQL dialect. Fixes #59036 and #59037.


This commit fixes the issues:
 https://github.com/ClickHouse/ClickHouse/issues/59036
 https://github.com/ClickHouse/ClickHouse/issues/59037

both issues are same reason, the input query exceed the max_query_size, so the condition isEnd() of token is not meet and cause the assertion failure

kashwy committed:
Fix_kql_issue_found_by_wingfuzz

This commit fix the issues:
 https://github.com/ClickHouse/ClickHouse/issues/59036
 https://github.com/ClickHouse/ClickHouse/issues/59037

both issues are same reason, the input query exceed the max_query_size,
so the condition isEnd() of token is not meet and cause the assertion failure

robot-ch-test-poll4 commented:
<!-- automatic status comment for PR #59626 from kashwy/ClickHouse:fix_kql_issue_found_by_wingfuzz -->
*This is an automated comment for commit 84f09d0d50afadae73b7b65df6e0243ece34bfe3 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59626/84f09d0d50afadae73b7b65df6e0243ece34bfe3/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td><a href="https://github.com/ClickHouse/ClickHouse/blob/master/.github/PULL_REQUEST_TEMPLATE.md?plain=1">❌ failure</a></td></tr>
<tbody>
</table>

antaljanosbenjamin commented on src/Parsers/Kusto/ParserKQLQuery.cpp:
```diff
@@ -373,7 +373,10 @@ bool ParserKQLQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
     uint16_t bracket_count = 0;
 
-    while (!pos->isEnd() && pos->type != TokenType::Semicolon)
+    while ((pos->type < TokenType::EndOfStream ||
+            pos->type == TokenType::ErrorSingleExclamationMark || // enable kql negative operators
+            String(pos->begin, pos->end) == "~") &&  // enable kql Case-Sensitive operators
+            pos->type != TokenType::Semicolon)
     {
```
```suggestion
    while ((pos.isValid() |
            pos->type == TokenType::ErrorSingleExclamationMark || // enable kql negative operators
            std::string_view(pos->begin, pos->end) == "~") &&  // enable kql Case-Sensitive operators
            pos->type != TokenType::Semicolon)
    {
```

antaljanosbenjamin reviewed (commented):
Apart from these small changes I think the PR looks fine. I haven't checked the tests throrougly, but I assume it is coming from a good source and/or validated with other KQL parsers.

kashwy committed:
fix_kql_issue_found_by_wingfuzz: use isValid instead of TokenType::EndOfStream

antaljanosbenjamin commented:
I edited the PR description to hopefully make the PR check happy.

kashwy commented:
> I edited the PR description to hopefully make the PR check happy.

thanks

yakov-olkhovskiy commented:
I think we also need tests for these bugs

kashwy commented:
> I think we also need test for this bug

this can be tested manually without  "--multiquery", otherwise the ErrorMaxQuerySizeExceeded is not triggered

kashwy commented:
Hi @antaljanosbenjamin , can you do the Wing Fuzz test to verify the issue:

https://github.com/ClickHouse/ClickHouse/issues/59036
https://github.com/ClickHouse/ClickHouse/issues/59037

thanks

kashwy committed:
fix_kql_issue_found_by_wingfuzz: make functional test result consist

kashwy commented:
> I can try to check the wingfuzz tests, but it is not clear to me how to reproduce the issue. I have also other tasks, so it might take some time.
> 
> In the meantime, I think the failing KQL test should be fixed. If the row order is random, we should either remove the test or make the row order deterministic.

yes, I have the test result updated.
for the wingfuzz tests, I think there is deadline for it, can you do  it ASAP ? 
thanks

kashwy committed:
fix_kql_issue_found_by_wingfuzz: update test case for makeseries

kashwy committed:
fix_kql_issue_found_by_wingfuzz: disable makeseries

kashwy commented:
@antaljanosbenjamin, the failures are not related with this PR now

kashwy commented:
@antaljanosbenjamin 
is there any news about the WING Fuzz test?

Best Regards

alexey-milovidov commented on src/Parsers/Kusto/ParserKQLQuery.cpp:
```diff
@@ -373,9 +373,9 @@ bool ParserKQLQuery::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
 
     uint16_t bracket_count = 0;
 
-    while ((pos->type < TokenType::EndOfStream ||
```
Ok, but it works exactly the same.

kashwy replied:
before we use  pos->isEnd() to validate, but if  when reach max_query_size,  the iterator already passed the end

alexey-milovidov replied:
The code looks the same: `bool isValid() { return get().type < TokenType::EndOfStream; }`

kashwy replied:
oh, yes, changed according to @antaljanosbenjamin 's suggestion

antaljanosbenjamin replied:
Because `pos->isEnd()` would be `false` in case of `TokenType::ErrorMaxQuerySizeExceeded`, thus `!pos->isEnd()` wouldn't terminate the loop. Meanwhile `pos->isValid()` will be false in case of `TokenType::ErrorMaxQuerySizeExceeded`, thus stopping the loop. Do I miss something?

alexey-milovidov commented:
Do you mean that the `makeseries` function contains a bug? Then, we should remove or fix it.

antaljanosbenjamin commented:
> Do you mean that the makeseries function contains a bug? Then, we should remove or fix it.

Actually this might a general issue in KQL parsers. If I search for `isEnd()` in `src/Parsers` most of the matches are KQL parsers:
```
 grep -Rnwl './src/Parsers' -e 'isEnd()'
./src/Parsers/parseQuery.cpp
./src/Parsers/examples/lexer.cpp
./src/Parsers/Kusto/ParserKQLMakeSeries.cpp
./src/Parsers/Kusto/ParserKQLMVExpand.cpp
./src/Parsers/Kusto/parseKQLQuery.cpp
./src/Parsers/Kusto/ParserKQLSort.cpp
./src/Parsers/Kusto/ParserKQLStatement.cpp
./src/Parsers/Kusto/ParserKQLOperators.cpp
./src/Parsers/Kusto/ParserKQLExtend.cpp
./src/Parsers/Kusto/ParserKQLSummarize.cpp
./src/Parsers/Kusto/KustoFunctions/KQLDateTimeFunctions.cpp
./src/Parsers/Kusto/KustoFunctions/KQLDataTypeFunctions.cpp
./src/Parsers/Kusto/KustoFunctions/KQLStringFunctions.cpp
./src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp
./src/Parsers/Kusto/ParserKQLQuery.cpp
./src/Parsers/PRQL/ParserPRQLQuery.cpp
./src/Parsers/toOneLineQuery.cpp
./src/Parsers/Lexer.h
./src/Parsers/TokenIterator.cpp
./src/Parsers/queryNormalization.cpp
./src/Parsers/obfuscateQueries.cpp
./src/Parsers/fuzzers/lexer_fuzzer.cpp
```

This might mean that all of these checks in Kusto has to be addresssed:

```
grep -Rnw './src/Parsers/Kusto' -e 'isEnd()'
./src/Parsers/Kusto/ParserKQLMakeSeries.cpp:42:    while (!pos->isEnd() && pos->type != TokenType::PipeMark && pos->type != TokenType::Semicolon)
./src/Parsers/Kusto/ParserKQLMakeSeries.cpp:99:    while (!pos->isEnd() && pos->type != TokenType::PipeMark && pos->type != TokenType::Semicolon)
./src/Parsers/Kusto/ParserKQLMakeSeries.cpp:178:        while (!pos->isEnd())
./src/Parsers/Kusto/ParserKQLMakeSeries.cpp:204:        while (!pos->isEnd())
./src/Parsers/Kusto/ParserKQLMVExpand.cpp:52:    while (!pos->isEnd() && pos->type != TokenType::PipeMark && pos->type != TokenType::Semicolon)
./src/Parsers/Kusto/ParserKQLMVExpand.cpp:128:        if (!pos->isEnd())
./src/Parsers/Kusto/ParserKQLMVExpand.cpp:130:        if (pos->isEnd() || pos->type == TokenType::PipeMark || pos->type == TokenType::Semicolon)
./src/Parsers/Kusto/parseKQLQuery.cpp:250:    for (TokenIterator it = begin; !it->isEnd(); ++it)
./src/Parsers/Kusto/parseKQLQuery.cpp:332:    if (token_iterator->isEnd()
./src/Parsers/Kusto/parseKQLQuery.cpp:406:    if (!token_iterator->isEnd()
./src/Parsers/Kusto/parseKQLQuery.cpp:424:        && !token_iterator->isEnd())
./src/Parsers/Kusto/ParserKQLSort.cpp:27:    while (!new_pos->isEnd() && new_pos->type != TokenType::PipeMark && new_pos->type != TokenType::Semicolon)
./src/Parsers/Kusto/ParserKQLStatement.cpp:82:            while (!pos->isEnd())
./src/Parsers/Kusto/ParserKQLOperators.cpp:151:    while (!token_pos->isEnd() && token_pos->type != TokenType::PipeMark && token_pos->type != TokenType::Semicolon)
./src/Parsers/Kusto/ParserKQLOperators.cpp:220:    while (!token_pos->isEnd() && token_pos->type != DB::TokenType::PipeMark && token_pos->type != DB::TokenType::Semicolon)
./src/Parsers/Kusto/ParserKQLOperators.cpp:332:    if (!pos->isEnd() && pos->type != TokenType::PipeMark && pos->type != TokenType::Semicolon)
./src/Parsers/Kusto/ParserKQLOperators.cpp:342:            if (pos->isEnd() || pos->type == TokenType::PipeMark || pos->type == TokenType::Semicolon)
./src/Parsers/Kusto/ParserKQLOperators.cpp:349:            if (!pos->isEnd() && pos->type != TokenType::PipeMark && pos->type != TokenType::Semicolon)
./src/Parsers/Kusto/ParserKQLOperators.cpp:363:        if (!pos->isEnd() && pos->type != TokenType::PipeMark && pos->type != TokenType::Semicolon)
./src/Parsers/Kusto/ParserKQLExtend.cpp:47:    while (!npos->isEnd())
./src/Parsers/Kusto/ParserKQLSummarize.cpp:123:                if (equal_pos->isEnd() || equal_pos->type == TokenType::Comma || equal_pos->type == TokenType::Semicolon
./src/Parsers/Kusto/ParserKQLSummarize.cpp:154:    while (!pos->isEnd() && pos->type != TokenType::PipeMark && pos->type != TokenType::Semicolon)
./src/Parsers/Kusto/KustoFunctions/KQLDateTimeFunctions.cpp:524:    while (!pos->isEnd() && pos->type != TokenType::ClosingRoundBracket)
./src/Parsers/Kusto/KustoFunctions/KQLDateTimeFunctions.cpp:591:    while (!pos->isEnd() && pos->type != TokenType::ClosingRoundBracket)
./src/Parsers/Kusto/KustoFunctions/KQLDataTypeFunctions.cpp:54:        while (!pos->isEnd() && pos->type != TokenType::PipeMark && pos->type != TokenType::Semicolon)
./src/Parsers/Kusto/KustoFunctions/KQLDataTypeFunctions.cpp:80:    while (!pos->isEnd() && pos->type != TokenType::ClosingRoundBracket)
./src/Parsers/Kusto/KustoFunctions/KQLDataTypeFunctions.cpp:120:        while (!pos->isEnd() && pos->type != TokenType::PipeMark && pos->type != TokenType::Semicolon)
./src/Parsers/Kusto/KustoFunctions/KQLStringFunctions.cpp:593:    while (!pos->isEnd() && pos->type != TokenType::Semicolon && pos->type != TokenType::ClosingRoundBracket)
./src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp:76:    while (!pos->isEnd() && pos->type != TokenType::PipeMark && pos->type != TokenType::Semicolon)
./src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp:151:    if (pos->isEnd() || pos->type == TokenType::PipeMark || pos->type == TokenType::Semicolon)
./src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp:155:    while (!pos->isEnd() && pos->type != TokenType::PipeMark && pos->type != TokenType::Semicolon)
./src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp:194:                    while (!pos->isEnd() && pos->type != TokenType::ClosingSquareBracket)
./src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp:249:    while (!pos->isEnd() && (!scopes.empty() || (pos->type != DB::TokenType::Comma && pos->type != DB::TokenType::ClosingRoundBracket)))
./src/Parsers/Kusto/KustoFunctions/IParserKQLFunction.cpp:360:        while (!pos->isEnd() && pos->type != TokenType::ClosingSquareBracket)
./src/Parsers/Kusto/ParserKQLQuery.cpp:133:    while (!end->isEnd() && end->type != TokenType::Semicolon)
./src/Parsers/Kusto/ParserKQLQuery.cpp:152:    while (!pos->isEnd() && pos->type != TokenType::Semicolon)
./src/Parsers/Kusto/ParserKQLQuery.cpp:376:    while (!pos->isEnd() && pos->type != TokenType::Semicolon)
```

I am not sure, but it is suspicious. I will try to play with them and find new bugs.

antaljanosbenjamin commented:
@kashwy if you want to help, you can also check out what happend when some of these `isEnd()` checks encounters `TokenType::ErrorMaxQuerySizeExceeded`.

kashwy commented:
> Do you mean that the `makeseries` function contains a bug? Then, we should remove or fix it.

will fixed with a separated PR

kashwy commented:
> @kashwy if you want to help, you can also check out what happend when some of these `isEnd()` checks encounters `TokenType::ErrorMaxQuerySizeExceeded`.

I can change all of them, although `TokenType::ErrorMaxQuerySizeExceeded`. will be screened out at the first place

antaljanosbenjamin commented:
> I can change all of them, although TokenType::ErrorMaxQuerySizeExceeded. will be screened out at the first place

I miss something in the logic then. Could you please clarify?

kashwy commented:
> > I can change all of them, although TokenType::ErrorMaxQuerySizeExceeded. will be screened out at the first place
> 
> I miss something in the logic then. Could you please clarify?

sure.
the current change is the entry point of kql parsing , if  encountered  TokenType::ErrorMaxQuerySizeExceeded, it will throw exception, and will not go to other parser

antaljanosbenjamin commented:
> the current change is the entry point of kql parsing , if encountered TokenType::ErrorMaxQuerySizeExceeded, it will throw exception, and will not go to other parser

But what if other parsers are started to parse a portion of the query? E.g.:
```
$ cat query.txt
SET dialect = 'kusto';
SET max_query_size = 16;
Customers | where Education contains 'degree'
$ ./clickhouse_fixed -n < query.txt 
Code: 1001. DB::Exception: std::bad_alloc: std::bad_alloc. (STD_EXCEPTION)
```

I think what happens here is:
1. `Customers` is parsed as a `table` operator => everything is fine
2. A `ParserKQLFilter` starts to parse the `Education contains 'degree'` part. In this part the query size limit is reached in `'degree'`, so there will be the `TokenType::ErrorMaxQuerySizeExceeded`, however `ParserKQLBase :: getExprFromToken` only checks `!pos->isEnd()`, this it passes the end of the string and we end up with a huuuuuuge string, that is why `std::bad_alloc` is thrown.

But there is another example:
```
$ cat query.txt
SET dialect = 'kusto';
SET max_query_size = 44;
Customers | where Education contains 'degree' | order by LastName
$ ./clickhouse_fixed -n < query.txt 
Code: 1001. DB::Exception: std::length_error: basic_string. (STD_EXCEPTION)
```
Here the same thing happens, but with `ParserKQLSort`. If you use the debug builds, it is obvious that this is wrong:

```
$ cat query.txt
SET dialect = 'kusto';
SET max_query_size = 44;
Customers | where Education contains 'degree' | order by LastName
$ ./clickhouse_fixed_debug -n < query.txt 
clickhouse_fixed_debug: /build/src/Parsers/TokenIterator.h:31: const Token &DB::Tokens::operator[](size_t): Assertion `index < data.size()' failed.
Aborted (core dumped)
```
I think this needs to be fixed. All of the `isEnd()` calls in KQL parsers are potential issues.

(BTW: I just realized that Alexey's comment is about the tests, but it directed my focus and I think we ended up with a better result).

kashwy commented:
> > the current change is the entry point of kql parsing , if encountered TokenType::ErrorMaxQuerySizeExceeded, it will throw exception, and will not go to other parser
> 
> But what if other parsers are started to parse a portion of the query? E.g.:
> 
> ```
> $ cat query.txt
> SET dialect = 'kusto';
> SET max_query_size = 16;
> Customers | where Education contains 'degree'
> $ ./clickhouse_fixed -n < query.txt 
> Code: 1001. DB::Exception: std::bad_alloc: std::bad_alloc. (STD_EXCEPTION)
> ```
> 
> I think what happens here is:
> 
> 1. `Customers` is parsed as a `table` operator => everything is fine
> 2. A `ParserKQLFilter` starts to parse the `Education contains 'degree'` part. In this part the query size limit is reached in `'degree'`, so there will be the `TokenType::ErrorMaxQuerySizeExceeded`, however `ParserKQLBase :: getExprFromToken` only checks `!pos->isEnd()`, this it passes the end of the string and we end up with a huuuuuuge string, that is why `std::bad_alloc` is thrown.
> 
> But there is another example:
> 
> ```
> $ cat query.txt
> SET dialect = 'kusto';
> SET max_query_size = 44;
> Customers | where Education contains 'degree' | order by LastName
> $ ./clickhouse_fixed -n < query.txt 
> Code: 1001. DB::Exception: std::length_error: basic_string. (STD_EXCEPTION)
> ```
> 
> Here the same thing happens, but with `ParserKQLSort`. If you use the debug builds, it is obvious that this is wrong:
> 
> ```
> $ cat query.txt
> SET dialect = 'kusto';
> SET max_query_size = 44;
> Customers | where Education contains 'degree' | order by LastName
> $ ./clickhouse_fixed_debug -n < query.txt 
> clickhouse_fixed_debug: /build/src/Parsers/TokenIterator.h:31: const Token &DB::Tokens::operator[](size_t): Assertion `index < data.size()' failed.
> Aborted (core dumped)
> ```
> 
> I think this needs to be fixed. All of the `isEnd()` calls in KQL parsers are potential issues.
> 
> (BTW: I just realized that Alexey's comment is about the tests, but it directed my focus and I think we ended up with a better result).

sure, I am fixing it

kashwy committed:
fix_kql_issue_found_by_wingfuzz:
 use isvalid() function to replace isEnd() function of TokenIterator to check the end of stream

kashwy commented:
@antaljanosbenjamin the issues are fixed:

```
SET dialect = 'kusto';
SET max_query_size = 16;
Customers | where Education contains 'degree'

Syntax error: failed at position 13 ('where'):
Customers | where Education contains 'degree'
Max query size exceeded: 'where'
```


```
SET dialect = 'kusto';
SET max_query_size = 44;
Customers | where Education contains 'degree' | order by LastName

Exception on client:
Code: 62. DB::Exception: Syntax error near contains. (SYNTAX_ERROR)
```

this is because the Max query size in the middle of the quote,



if  `SET max_query_size = 55;`

```
Customers | where Education contains 'degree' | order by LastName

Syntax error: failed at position 55 ('by'):

Customers | where Education contains 'degree' | order by LastName

Max query size exceeded: 'by'
```

kashwy committed:
fix_kql_issue_found_by_wingfuzz: add test case for max_query_size

kashwy committed:
fix_kql_issue_found_by_wingfuzz: fix AST structure

kashwy committed:
fix_kql_issue_found_by_wingfuzz: make sure the max query size test is in the dialect of kusto

kashwy committed:
fix_kql_issue_found_by_wingfuzz : restore max query size after test

yakov-olkhovskiy commented:
@kashwy there is a typo:
```
2024-02-19 07:12:09 /ClickHouse/src/Parsers/Kusto/ParserKQLQuery.cpp:645:5: error: use of undeclared identifier 'node_table_in_select_query_emlement'; did you mean 'node_table_in_select_query_element'?
2024-02-19 07:12:09   645 |     node_table_in_select_query_emlement->children.emplace_back(node_table_expr);
2024-02-19 07:12:09       |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2024-02-19 07:12:09       |     node_table_in_select_query_element
2024-02-19 07:12:09 /ClickHouse/src/Parsers/Kusto/ParserKQLQuery.cpp:642:12: note: 'node_table_in_select_query_element' declared here
2024-02-19 07:12:09   642 |     ASTPtr node_table_in_select_query_element = std::make_shared<ASTTablesInSelectQueryElement>();
2024-02-19 07:12:09       |            ^
2024-02-19 07:12:09 1 error generated.
```

kashwy committed:
fix_kql_issue_found_by_wingfuzz : fix typo

antaljanosbenjamin commented:
I think overall the PR is mergable, I don't see any possible dangerous issues.

However in the future (not in this PR though) this should be fixed:

> this is because the Max query size in the middle of the quote,

I think the reason here is because Kusto parsers sometimes written in a way that they don't have any other way to signal errors than throwing an exception, e.g.: `String KQLOperators::genHasAnyAllOpExpr(std::vector<String> & tokens, IParser::Pos & token_pos, String kql_op, String ch_op)`. Maybe they should return optional values and the simple parsing errors should be handled on higher levels. I am not sure, but the error message is too generic here. But again, this shouldn't be part of this PR, this PR aims to fix crashes, which it does in my opnion.

kashwy commented:
Hi @antaljanosbenjamin , there are some test failures ,but I did not find anything related with my commit, can you help to take a look ?

Thanks

kashwy commented:
> I think overall the PR is mergable, I don't see any possible dangerous issues.
> 
> However in the future (not in this PR though) this should be fixed:
> 
> > this is because the Max query size in the middle of the quote,
> 
> I think the reason here is because Kusto parsers sometimes written in a way that they don't have any other way to signal errors than throwing an exception, e.g.: `String KQLOperators::genHasAnyAllOpExpr(std::vector<String> & tokens, IParser::Pos & token_pos, String kql_op, String ch_op)`. Maybe they should return optional values and the simple parsing errors should be handled on higher levels. I am not sure, but the error message is too generic here. But again, this shouldn't be part of this PR, this PR aims to fix crashes, which it does in my opnion.

yeah, will do later

antaljanosbenjamin commented:
So the issue is we recently made some changes in master in order to enforce that queries are formatted and parsed properly. That caused a lot of noise in our CI. As this PR is related to parsers, I will merge a recent master to rerun the tests. So far I haven't seen anything KQL related, but I might be wrong because of the huge number of failed tests I haven't checked every test in details.

antaljanosbenjamin committed:
Merge remote-tracking branch 'origin/master' into fix_kql_issue_found_by_wingfuzz

kashwy commented:
> So the issue is we recently made some changes in master in order to enforce that queries are formatted and parsed properly. That caused a lot of noise in our CI. As this PR is related to parsers, I will merge a recent master to rerun the tests. So far I haven't seen anything KQL related, but I might be wrong because of the huge number of failed tests I haven't checked every test in details.

thank you very much.  I did not see errors related with kql. 
Can this PR  be merged ? 

Best regard

antaljanosbenjamin commented:
I have rerun the clang-tidy build, because that is required check. After that I think the PR can be merged.

kashwy commented:
> I have rerun the clang-tidy build, because that is required check. After that I think the PR can be merged.

thanks

antaljanosbenjamin commented:
Test failures:
- `Stress Test (ubsan)` #60349
- `00984_parser_stack_overflow` #60216
- `test_replicated_database/test.py::test_startup_without_zk` #60215
